json_configs/configuration.cpp

```diff
@@ -0,0 +1,6398 @@
+ *
+ * This file is part of Vega Strike.
+ *
+ * Vega Strike is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Vega Strike is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+
+#include <iostream>
+#include <exception>
+#include <boost/json.hpp>
+#include <boost/property_tree/json_parser.hpp>
+
+#include "configuration.h"
+
+#ifndef _USE_MATH_DEFINES
+#define _USE_MATH_DEFINES
+#endif
+#include <math.h>
+#include <vs_logging.h>
+#include <boost/format/format_fwd.hpp>
+#include <boost/program_options/errors.hpp>
+
+vega_config::Configuration::Configuration()
+= default;
+
+vega_config::Configuration::Configuration(const std::string& json_text) {
+    load_config(json_text);
+}
+
+vega_config::Configuration::Configuration(const boost::filesystem::path& config_file_path) {
+    load_config(config_file_path);
+}
+
+void vega_config::Configuration::load_config(const boost::filesystem::path& config_file_path) {
+    try {
+        std::ifstream config_file(config_file_path.string());
+        std::string str;
+        std::string file_contents;
+        while (std::getline(config_file, str)) {
+            file_contents += str;
+            file_contents.push_back('\n');
+        }
+        load_config(file_contents);
+    } catch (const std::exception& e) {
+        std::cerr << e.what() << std::endl;
+    }
+}
+
+void vega_config::Configuration::load_config(const std::string& json_text) {
+    try {
+        if (json_text.empty()) {
+            return;
+        }
+        boost::json::value json_value = boost::json::parse(json_text);
+        if (json_value.is_null()) {
+            return;
+        }
+        const boost::json::object & root_object = json_value.get_object();
+
+
+
+        const boost::json::value * advanced_value_ptr = root_object.if_contains("advanced");
+        if (advanced_value_ptr != nullptr) {
+            boost::json::object advanced_object = advanced_value_ptr->get_object();
+        }
+
+
+        const boost::json::value * ai_value_ptr = root_object.if_contains("ai");
+        if (ai_value_ptr != nullptr) {
+            boost::json::object ai_object = ai_value_ptr->get_object();
+            const boost::json::value * afterburn_to_no_enemies_value_ptr = ai_object.if_contains("afterburn_to_no_enemies");
+            if (afterburn_to_no_enemies_value_ptr != nullptr) {
+                ai.afterburn_to_no_enemies = boost::json::value_to<bool>(*afterburn_to_no_enemies_value_ptr);
+            }
```

Comment:
there's enough duplication here that perhaps we should make a C Macro for each type?
```c
#define GET_BOOL_VALUE(name, name2, target) \
{ /* wrap as a block to make a namespace for the macro */ \
const boost::json::value* dest = root_object.if_container(name);  \
if (dest != nullptr)  { \
   boost::json::value* dest2 = dest->get_object(); \
   const boost::json::value* dest3 = dest2.if_contains(name2); \
    if (dest3 != nullptr) { \
        target = boost::json::value_to<bool>(*dest3); \
    } \
}
```

Reply:
The feedback I've gotten on C macros in places like StackOverflow has been pretty consistent: No C macros. No more. They aren't needed with modern C++ constructs available.

With that said, perhaps a set of inline functions would do. Or something like that.

Reply:
Two comments:
1. This is an auto-generated file. We should mark it, so people won't be tempted to refactor the obviously inefficient code.
2. It's auto-generated. There's really no reason to refactor this.

Reply:
@BenjamenMeyer 
Just in case you forgot or missed it, this is based on a utility I wrote outside of the Vegastrike project. 
It takes a nested JSON file and auto-generates a cpp and header files to store its values.

This means that any changes to the configuration JSON can be easily reflected in the code without additional work. Given the size of the vegastrike.config file, it seemed like a good idea.

Reply:
Can we add a header to it via the utility then?
Most tooling something like the following for autogenerated files:
````
// DO NOT EDIT
// Autogenerated by tool X on 2025-05-12 at 12:36 AM UTC
```

Reply:
> The feedback I've gotten on C macros in places like StackOverflow has been pretty consistent: No C macros. No more. They aren't needed with modern C++ constructs available.
> 
> With that said, perhaps a set of inline functions would do. Or something like that.

That's because most on Stackoverflow don't understand C Macros and many abuse them improperly. There's a place for them. C++ inline functions can't quite do everything that a C Macro can. They do cover most cases (90%?); but not all, and no Templates don't help cover the other 10%.

Most use them more than they realize too.

That said, I don't care if it's a Macro or inline function just would like to see it cleaned up.

Reply:
@BenjamenMeyer I have added a header to each of the files noting that they are autogenerated.